### DISCUSSION TITLE
Implement AgX as a tonemapper

### DISCUSSION THREAD

**Discussion by ettiSurreal** [godotengine/godot-proposals#7545] [3D]
Very recently, Blender added Eary's AgX (improved version over the original config, if we do implement AgX into Godot, I think this should be the version to implement: https://github.com/EaryChow/AgX) as a new view transform in it's color management.

In Godot there are currently 4 options: Linear, Reinhard, Filmic and ACES. Unfortunately all of these have their issues. Reinhard and Filmic suffer from problems like the notorious six (https://github.com/godotengine/godot-proposals/issues/6096) and ACES is... problematic to say the least. 
AgX fixes the issues Filmic had, and is all around a great tonemapper.

Godot doesn't utilize OCIO, so I don't know how one could implement a new tonemapper, but this discussion isn't about bringing OCIO support to the engine.

Related:
https://github.com/godotengine/godot-proposals/issues/7263

**Comment by Calinou** [comment]
Given the complexity of the OCIO file, this likely requires baking a LUT and using that LUT during tonemapping, which has the same binary size concerns mentioned in https://github.com/godotengine/godot-proposals/issues/7263.

This would steer me towards implementing a Custom tonemapping mode which allows you to specify your own 3D LUT to be used *before* color correction (and before tonemapping). LUTs for tonemapping can then be provided on the asset library.

**Reply by s-ilent** [reply]
I think that makes a lot of sense, but I'm worried a full implementation of OCIO and pipeline for importing them as LUTs might be very heavy duty for a lot of Godot users. As Blender has moved to use AgX as a default (through an OCIO profile, to be fair), I think it's also worth considering adding in an option for minimal AgX as an inbuilt tonemapper option for Godot for easy Blender parity. 

https://iolite-engine.com/blog_posts/minimal_agx_implementation

**Reply by Calinou** [reply]
Thanks to the code you linked, I have a working proof of concept with AGX and AGX Punchy: https://github.com/Calinou/godot/tree/tonemap-add-agx

**Testing project:** [tonemap_color_correction.zip](https://github.com/godotengine/godot-proposals/files/13805380/tonemap_color_correction.zip)

However, I'm not sure how the whitepoint property is supposed to be implemented. The blog post doesn't mention it, and any other references seem to imply the use of OpenColorIO or a standard D65 whitepoint (while we want to be able to adjust it here).

Also, if I want to open a PR, I'll have to contact the author to use the code under the MIT license (since the blog post or Shadertoy don't mention a license).

## Preview

### Linear

![Screenshot_20240101_180950 webp](https://github.com/godotengine/godot-proposals/assets/180032/de04f29d-b19f-468b-b6ed-7f06f8e26d1d)

### Filmic

![Screenshot_20240101_180945 webp](https://github.com/godotengine/godot-proposals/assets/180032/8c58e2e9-9036-4b12-b06c-9a09fa52ef3a)

### ACES

![Screenshot_20240101_180954 webp](https://github.com/godotengine/godot-proposals/assets/180032/e505e25b-fffc-49fa-a349-7ca5d719ef4d)

### AGX

*Whitepoint is ignored here.*

![Screenshot_20240101_180958 webp](https://github.com/godotengine/godot-proposals/assets/180032/f32512dc-159c-48c1-817e-27b2e5ab5147)

### AGX Punchy

*Whitepoint is ignored here.*

![Screenshot_20240101_181002 webp](https://github.com/godotengine/godot-proposals/assets/180032/f731666c-07d8-4cda-b695-5e084fe5d659)

**Reply by Calinou** [reply]
Pull request opened: https://github.com/godotengine/godot/pull/87260

**Comment by statixVFX** [comment]
+1 for AgX support.

It solves a very important problem in computer graphics when it comes to color reproduction with high saturation (the collapsing 6 colours)

AgX vs rec709 (which has the same 6 collapsing colours as srgb, filmic srgb and aces)
![228655449-0b9b5e7b-e962-400f-bfb5-56c104bc7cd9](https://github.com/godotengine/godot-proposals/assets/5736513/2ba342c8-1f92-419c-9239-8a032b2689cc)

**Reply by redhoot-dev** [reply]
+1 for AgX, its been phenomenal in Blender and other places. 

It solves the problem described above of the collapsing 6 colors witch have plagued computer graphics for ages. 

Look at any modern AAA game, and you still see colors skew harshly towards yellow, magenta when over-exposing. 
This is very evident in areas of high contrast, like weapon-shot lighting, FX, impacts or just harsh lighting.

**Reply by DigitallyTailored** [reply]
Not sure if this is the right place to comment but how do we access `Agx Punchy`?

**Reply by Calinou** [reply]
> Not sure if this is the right place to comment but how do we access `Agx Punchy`?

AGX Punchy was removed from the final implementation that was merged in 4.4. You can get a similar result by enabling **Adjustments** and increasing Contrast/Saturation (try setting both to `1.3`).

**Comment by akien-mga** [comment]
Implemented with https://github.com/godotengine/godot/pull/87260.
